src/main/scala/mesosphere/marathon/upgrade/DeploymentPlanReverter.scala

```diff
@@ -0,0 +1,165 @@
+      existingGroup.copy(dependencies = existingGroup.dependencies ++ oldGroup.dependencies)
+    }
+
+    def revertDependencyChanges(oldGroup: Group, newGroup: Group)(group: Group): Group = {
+      val removedDependencies = oldGroup.dependencies -- newGroup.dependencies
+      val addedDependencies = newGroup.dependencies -- oldGroup.dependencies
+
+      if (removedDependencies.nonEmpty || addedDependencies.nonEmpty) {
+        if (log.isDebugEnabled)
+          log.debug(
+            s"revert dependency changes in group ${oldGroup.id}, " +
+              s"readding removed {${removedDependencies.mkString(", ")}}, " +
+              s"removing added {${addedDependencies.mkString(", ")}}")
+
+        group.copy(dependencies = group.dependencies ++ removedDependencies -- addedDependencies)
+      }
+      else {
+        // common case, unchanged
+        group
+      }
+    }
+
+    def revertGroupAddition(result: Group, newGroup: Group): Group = {
+      // We know that all group removals for groups inside of this group have been
+      // performed before. If there are no conflicts with other deployments, we could
+      // just remove the group. Unfortunately, another deployment could have
+      // added a new sub group or sub app definition. In this case, we should not remove
+      // this group.
+
+      def normalized(group: Group): Group = group.withNormalizedVersion.withoutChildren
+      def isGroupUnchanged(group: Group): Boolean =
+        !group.containsAppsOrGroups && normalized(group) == normalized(newGroup)
+
+      result.group(newGroup.id) match {
+        case Some(unchanged) if isGroupUnchanged(unchanged) =>
+
+          log.debug("remove unchanged group {}", unchanged.id)
+          result.remove(unchanged.id)
+        case _ => // group still contains apps/groups or the dependencies have changed
+          if (log.isDebugEnabled)
+            log.debug(s"group ${newGroup.id} has changed. " +
+              s"Removed added dependencies ${newGroup.dependencies.mkString(", ")}")
+          result.update(
+            newGroup.id,
+            group => group.copy(dependencies = group.dependencies -- newGroup.dependencies),
+            Timestamp.now())
+      }
+    }
+
+    // sort groups so that inner groups are processed first
+    val sortedGroupChanges = groupChanges.sortWith {
+      case (change1, change2) =>
+        // both groups are supposed to have the same path id (if there are any)
+        def pathId(change: (Option[Group], Option[Group])): PathId = {
+          Seq(change._1, change._2).flatten.map(_.id).headOption.getOrElse(PathId.empty)
+        }
+
+        pathId(change1) > pathId(change2)
+    }
+
+    sortedGroupChanges.foldLeft(group) {
+      case (result, groupUpdate) =>
+        groupUpdate match {
+          case (Some(oldGroup), None) =>
+            result.update(oldGroup.id, revertGroupRemoval(oldGroup), Timestamp.now())
```

Comment:
This will create different Timestamps for different changes. Please use one Timestamp for the whole change.